### PR TITLE
Knockback improve

### PR DESCRIPTION
**Pull request by SaithTime**
This is a big improvement in knockback, especially useful for practice servers, but it can be useful for Pvp oriented servers.

- First of all, it improves the management of knockback values.
- Improved knockback command with clickable interface (I thought I was going to die).
- Added knockback management for projectiles.
- Added an instant-interaction parameter, this will reduce the delay of the playinuseentity package used in Pvp.

I propose you this PR because you have in your hands the most powerful spigot possible for the creation of a practice server.

### CODE REVIEW

File: NachoSpigot-Server/src/main/java/me/elier/util/ClickableBuilder.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package me.elier.util;
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/me/elier/util/ClickableBuilder.java:1]: Move this to Nacho's package or your own

File: NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java [hunk 2]
```diff
@@ -1014,34 +1014,25 @@ public void attack(Entity entity) {
...
-                            if(!config.customKnockback) entity.g(-MathHelper.sin(this.yaw * 3.1415927F / 180.0F) * (float) i * 0.5F, 0.1D, (double) (MathHelper.cos(this.yaw * 3.1415927F / 180.0F) * (float) i * 0.5F)); else
+                            KnockbackProfile profile = (entity.getKnockbackProfile() == null) ?
+                                    Nacho.get().getKnockbackConfig().getCurrentKb() : entity.getKnockbackProfile();
                             entity.g(
-                                    (-MathHelper.sin(this.yaw * 3.1415927F / 180.0F) * (float) i * config.knockbackExtraHorizontal),
-                                    config.knockbackExtraVertical,
-                                    (MathHelper.cos(this.yaw * 3.1415927F / 180.0F) * (float) i * config.knockbackExtraHorizontal));
+                                    (-FastMath.sin(this.yaw * Math.PI / 180.0D) * i * profile.getExtraHorizontal()),
+                                   profile.getExtraVertical(),
+                                    (FastMath.cos(this.yaw * Math.PI / 180.0D) * i * profile.getExtraHorizontal()));
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1022]:
```suggestion
                                    (-MathHelper.sin(this.yaw * Math.PI / 180.0D) * i * profile.getExtraHorizontal()),
                                   profile.getExtraVertical(),
                                    (MathHelper.cos(this.yaw * Math.PI / 180.0D) * i * profile.getExtraHorizontal()));
```
See https://github.com/CobbleSword/NachoSpigot/pull/231#issuecomment-968184311

File: NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java [hunk 3]
```diff
@@ -1014,34 +1014,25 @@ public void attack(Entity entity) {
...
 
                             if (event.isCancelled()) {
-                                cancelled = true;
-                            } else if (!velocity.equals(event.getVelocity())) {
                                 player.setVelocity(event.getVelocity());
-                            }
-
-                            if (!cancelled) {
-                                ( (EntityPlayer) entity ).playerConnection.sendPacket( new PacketPlayOutEntityVelocity( entity ) );
+                                ((EntityPlayer) entity).playerConnection.sendPacket(new PacketPlayOutEntityVelocity(entity));
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1035]: Why is this removed?

**Review by Elierrr — CHANGES_REQUESTED** [review]
Everything else LGTM, also if you wish you can put the new classes in your own package.

File: NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java (hunk 3 above)
Comment (Lucaskyy) [reply NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1035]: It seems that this skips the cancellation of the player velocity event, not sure why.
@SaithTime Could you explain?

Comment (SaithTime) [reply NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1035]: t's just a cleanup, the event is not removed

Comment (Elierrr) [reply NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1035]: > t's just a cleanup, the event is not removed

The default functionality of the event only runs while it is canceled in this code.

Comment (SaithTime) [reply NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1035]: Just a mistake on my part ^^

File: NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java [hunk 4]
```diff
@@ -1014,34 +1014,25 @@ public void attack(Entity entity) {
...
-
-                            PlayerVelocityEvent event = new PlayerVelocityEvent(player, velocity.clone());
+                            PlayerVelocityEvent event = new PlayerVelocityEvent(player, new Vector(entity.motX, entity.motY, entity.motZ));
                             world.getServer().getPluginManager().callEvent(event);
 
-                            if (event.isCancelled()) {
-                                cancelled = true;
-                            } else if (!velocity.equals(event.getVelocity())) {
+                            if (!event.isCancelled()) {
                                 player.setVelocity(event.getVelocity());
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1034]:
```suggestion
                                if (!velocity.equals(event.getVelocity())) player.setVelocity(event.getVelocity());
```

**Review by Elierrr — CHANGES_REQUESTED** [review]

File: NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java (hunk 4 above)
Comment (Lucaskyy) [reply NachoSpigot-Server/src/main/java/net/minecraft/server/EntityHuman.java:1034]: Shouldn't the `isCancelled` be checked too?

**Comment by Lucaskyy** [comment]
@SaithTime Hopefully there are no conflicts since I migrated the knockback config to static for you to make it the same as the NachoConfig.

**Comment by Lucaskyy** [comment]
I'll fix the conflict.

**Review by Elierrr — APPROVED** [review]
LGTM

**Comment by Lucaskyy** [comment]
Alright, should be all good now. Will test it now before I merge it :)

**Comment by Lucaskyy** [comment]
Just tested it and everything works 👍 
Thank you for this PR @SaithTime and thanks @Elierrr for helping/reviewing!

**Pull request merged**